Add ability to prevent editing blocks using useBlockEditingMode()

noisysocks opened the pull request:
## What, why, how
Alternative to https://github.com/WordPress/gutenberg/pull/50082.
Required to implement https://github.com/WordPress/gutenberg/pull/50857.
Part of https://github.com/WordPress/gutenberg/issues/44461.

This PR adds a new API to `block-editor` that lets one restrict the editing UI of a block and its children. It's based off of the suggestion by @youknowriad in https://github.com/WordPress/gutenberg/pull/50082#issuecomment-1529406156.

A new hook is added called `useBlockEditingMode( clientId, mode )`:

```js
function PostTitleEdit( { clientId } ) {
	useBlockEditingMode( clientId, 'disabled' );
	return (
		...
	);
}
```

 A block can call this in its `edit` function (or anywhere, really) with `mode` set to one of the three options:

- `'disabled'` - This prevents editing the block entirely. The block cannot be selected or interacted with. It uses the same `pointer-events: disabled` mechanism that is used when `templateLock` is set to `'contentOnly'`.
- `'contentOnly'` - This allows editing the block but removes most UI shown in the block toolbar so that only content related buttons (e.g. formatting) are shown. The block cannot be moved, deleted, etc. It's similar to the treatment received by a content block when `templateLock` is set to `'contentOnly'`.
- `'default'` - Default. Block is fully editable, moveable, removable, etc.

The block editing mode is inherited from the parent all the way up to `rootBlockEditingMode` which is specified in the editor settings. This lets one disable the entire editor except for certain blocks which is what is needed to implement https://github.com/WordPress/gutenberg/pull/49980.

The same hook can be used to return the current editing mode:

```js
const blockEditingMode = useBlockEditingMode( clientId );
```

This hook is a convenience wrapper for the `getBlockEditingMode( clientId )` selector and `setBlockEditingMode( clientId, mode )` action. Both exist in the `block-editor` store.

## Alternatives considered 

See https://github.com/WordPress/gutenberg/pull/50082 for my first approach at this.

The original idea suggested by @youknowriad in https://github.com/WordPress/gutenberg/pull/50082#issuecomment-1529406156 was to wrap blocks in a component that uses context to disable its children e.g. `<DisableBlocksProvider>`. This is not sufficient for our needs though as we need to be able to customise the UI in `BlockTools` and `BlockToolbar` which sits outside of the `BlockListBlock` component tree. We therefore need to be able to look up a block's editing mode by its client ID. Because the editing mode is inherited from the block's parent, this map needs to exist within the block editor store where we have access to the block tree.

## Testing Instructions
1. Apply the patch below by copying it and running `pbpaste | git apply`.

```diff
diff --git a/packages/edit-site/src/components/block-editor/index.js b/packages/edit-site/src/components/block-editor/index.js
index cc5e7c8d92..f97599a9b5 100644
--- a/packages/edit-site/src/components/block-editor/index.js
+++ b/packages/edit-site/src/components/block-editor/index.js
@@ -39,7 +39,9 @@ import EditorCanvas from './editor-canvas';
 import { unlock } from '../../private-apis';
 import EditorCanvasContainer from '../editor-canvas-container';
 
-const { ExperimentalBlockEditorProvider } = unlock( blockEditorPrivateApis );
+const { ExperimentalBlockEditorProvider, useBlockEditingMode } = unlock(
+	blockEditorPrivateApis
+);
 
 const LAYOUT = {
 	type: 'default',
@@ -47,6 +49,11 @@ const LAYOUT = {
 	alignments: [],
 };
 
+function SetRootBlockEditingMode( { mode } ) {
+	useBlockEditingMode( mode );
+	return null;
+}
+
 export default function BlockEditor() {
 	const { setIsInserterOpened } = useDispatch( editSiteStore );
 	const { storedSettings, templateType, canvasMode } = useSelect(
@@ -162,6 +169,7 @@ export default function BlockEditor() {
 			onChange={ onChange }
 			useSubRegistry={ false }
 		>
+			<SetRootBlockEditingMode mode="disabled" />
 			<TemplatePartConverter />
 			<SidebarInspectorFill>
 				<BlockInspector />
diff --git a/packages/edit-site/src/hooks/block-editing-mode.js b/packages/edit-site/src/hooks/block-editing-mode.js
new file mode 100644
index 0000000000..24d63f122f
--- /dev/null
+++ b/packages/edit-site/src/hooks/block-editing-mode.js
@@ -0,0 +1,50 @@
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { addFilter } from '@wordpress/hooks';
+import {
+	store as blockEditorStore,
+	privateApis as blockEditorPrivateApis,
+} from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../private-apis';
+
+const { useBlockEditingMode } = unlock( blockEditorPrivateApis );
+
+export const withBlockEditingMode = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const mode = useSelect( ( select ) => {
+			if (
+				[
+					'core/post-title',
+					'core/post-featured-image',
+					'core/post-content',
+				].includes( props.name )
+			) {
+				return 'contentOnly';
+			}
+			if (
+				select( blockEditorStore ).getBlockParentsByBlockName(
+					props.clientId,
+					'core/post-content'
+				).length
+			) {
+				return 'default';
+			}
+		} );
+		useBlockEditingMode( mode );
+		return <BlockEdit { ...props } />;
+	},
+	'withBlockEditingMode'
+);
+
+addFilter(
+	'editor.BlockEdit',
+	'core/edit-site/block-editing-mode',
+	withBlockEditingMode
+);
diff --git a/packages/edit-site/src/hooks/index.js b/packages/edit-site/src/hooks/index.js
index 513634c55b..de9dc8c3f7 100644
--- a/packages/edit-site/src/hooks/index.js
+++ b/packages/edit-site/src/hooks/index.js
@@ -1,6 +1,7 @@
 /**
  * Internal dependencies
  */
+import './block-editing-mode';
 import './components';
 import './push-changes-to-global-styles';
 import './template-part-edit';

```

2. Navigate to Appearance → Editor → Pages and select a page to edit.
3. All blocks should be locked except for the Post Title, Post Featured Image and Post Content blocks.

To test the existing `contentOnly` functionality for regressions:
1. Add a pattern that has a content lock, e.g. [this one](https://gist.github.com/richtabor/ddeea41ced691721318649bea8ce9db8).
2. Edit a post, page, or template.
3. Insert that pattern.

noisysocks committed:
Add ability to prevent editing blocks using useBlockEditingMode()

noisysocks commented:
WDYT of this alternative approach @youknowriad @talldan? If we like it I can add tests and docs.

noisysocks commented on packages/block-editor/src/components/block-list/content.scss:
```diff
@@ -161,13 +161,13 @@
+	pointer-events: initial;
+
+	&.is-editing-disabled {
 		pointer-events: none;
```
We can't use the `Disabled` component / the `inert` HTML attribute because that completely disables all descendants of the node. We want _some_ children (e.g. the Post Title, Post Content) blocks to still be enabled.

ramonjd replied:
It might be nice to have some opacity at play here, e.g., lower opacity when `&.is-editing-disabled` and full opacity for `:hover`. 

It might also be nice if this was a good idea 😄

noisysocks replied:
The problem with using `opactiy` (I tried it) is that it compounds. So e.g. if the opacity is 20% then a block that's 4 levels deep becomes `50% ^ 4 = 6.25%` opacity. I couldn't figure out a good solution to that.

noisysocks replied:
I think let's just start with this UI (the one we'll implement in https://github.com/WordPress/gutenberg/pull/49980) and iterate.

noisysocks commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -234,10 +233,11 @@ function BlockListBlock( {
 		className: classnames(
 			{
-				'is-content-locked': isContentLocking,
+				'is-editing-disabled':
+					blockEditingMode === 'disabled' ||
+					( hasContentLockedParent && ! isContentBlock ),
```
Something I'm not sure about is whether or not to have `getEditingMode` contain the template lock logic as well so that components only need to call `getBlockEditingMode` instead of both `getBlockEditingMode` and `getTemplateLock`.

It makes the selector logic more complicated, but means that blocks/consumers only have to worry about calling one method and adjusting the UI accordingly.

ramonjd replied:
> It makes the selector logic more complicated, but means that blocks/consumers only have to worry about calling one method and adjusting the UI accordingly.

If the API goes public my vote is for this. Maybe it'll be make it easier to maintain too if the conditions are all in one spot?

ramonjd replied:
Also to note - if I remember correctly, there's also the potential for `getBlockEditingMode` to take over the work of (or even replace) `__unstableGetContentLockingParent`?

The while loops look very similar. 

From what I can see `__unstableGetContentLockingParent` is exclusively used to check if a block is locked by its "inert" parent - the parent's clientId is never used. 

In my mind, it would be handy to know whether a block is disabled with one call.

 but maybe we should avoid over-thinking/optimization altogether 😄 

It's a tricky one. I'm also thinking, is there nothing a few meaningfully-named functions or properties couldn't tell us?


```js
isBlockEditingDisabled(); // blockEditingMode === 'disabled'
isLockedByParent();  // !! __unstableGetContentLockingParent( clientId )
isContentEditingEnabled(); // blockEditingMode === 'contentOnly' || templateLock === 'contentOnly'
```

See `useBlockLock` (packages/block-editor/src/components/block-lock/use-block-lock.js), that returns these sorts of things.

I'm not sure if it's related, but I'm wondering also if we need to taken into account any "lock" instructions on the block's attributes. See `canEditBlock` (packages/block-editor/src/store/selectors.js).

noisysocks replied:
I've put the `contentOnly` template lock logic into `getBlockEditingMode` in one of my more recent commits. I think it's definitely nice having only one selector to call, and hopefully the JSDoc comments I help demystify it. What do you think?

Respecting `canEditBlock` is an interesting idea. I'll investigate. From memory `canEditBlock` is a bit weird and we only use it in the navigation block.

noisysocks replied:
OK so you can test what `canEditBlock` does yourself by selecting a Navigation block, open the settings menu, selecting _Lock_, then enabling _Lock editing_.

When you do this you'll see that you can select the Navigation block but that there's an overlay meaning you can't select its children or edit anything. This is different from either `'disabled'` which prevents section altogether or `'contentOnly'` which allows editing but hides toolbars, etc.

I suppose we could add a fourth mode called e.g. `'overlay'` to `getBlockEditingMode` which encapsulates this behaviour. I'm not really sure there's any benefit to doing this though. `getBlockEditingMode` is a private API so it's something we can change easily in the future if there's a need.

noisysocks commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -18,3 +18,19 @@ export function isBlockInterfaceHidden( state ) {
+
+export function getBlockEditingMode( state, clientId ) {
+	if ( ! clientId ) {
+		return state.settings.rootBlockEditingMode;
```
Need to mark this as private somehow.

ramonjd replied:
Also noting your idea to have some sort of data component that one could insert at the root-level of a components tree that sets the "root" editing mode. 

```jsx
		<ExperimentalBlockEditorProvider
			settings={ settings }
			value={ blocks }
			onInput={ onInput }
			onChange={ onChange }
			useSubRegistry={ false }
		>
                      <RootBlockEditingMode mode={ ... } />
```

noisysocks replied:
I made this change in a recent commit—check it out. I kept it so that `block-editor` just exports a hook, though, instead of a `RootBlockEditingMode` component. That feels right to me as I don't want to be too prescriptive of how the hook is called.

youknowriad reviewed (commented):
If this gives us the right solution, in terms of behavior, then I'm sold. I like this API and I can even see ourselves making it public.

Some thoughts/questions:

 - Will this have any impact on performance?
 - I know that we can't use `inert`/`useDisabled` for the "contentOnly" mode but maybe we can still consolidate this with `__unstableHasActiveBlockOverlayActive` (make the latter use the former maybe) and when the mode is "disabled", rely on `useDisabled`? (It doesn't really have to be here, but curious for your thoughts)
 - I like how this also replaces the "contentOnly" lock behavior (so not really a new mode or something)

noisysocks commented:
Thanks @ramonjd for jumping on a call with me to review this PR. Notes from that call:

- Change `useBlockEditingMode( clientId, mode )` → `useBlockEditingMode( mode )`. Look up `clientId` from context. That way can call this hook from top level instead of having a public `settings.rootBlockEditingMode`.
- Put the "is template content locked?" logic into `getBlockEditingMode` selector so that we're not repeating ourselves everywhere. Document that the selector will do this logic. Possibly should be two selectors? Possibly should be done as a follow up?
- List View and breadcrumbs need work. Disabled blocks shouldn't be shown there.

> Will this have any impact on performance?

Will double check but should be okay so long as the `getBlockEditingMode` selector is memoised. (Need to do this.)

> I know that we can't use inert/useDisabled for the "contentOnly" mode but maybe we can still consolidate this with __unstableHasActiveBlockOverlayActive (make the latter use the former maybe) and when the mode is "disabled", rely on useDisabled? (It doesn't really have to be here, but curious for your thoughts)

@ramonjd and I spoke about this. We could put `inert` on blocks where we know that all of the block's descendants are disabled. Maybe something worth exploring in a follow up.

> I like how this also replaces the "contentOnly" lock behavior (so not really a new mode or something)

To be clear: `templateLock = 'contentOnly'` still exists. Just adding a different mechanism to activate that same functionality.

I'll address all this feedback and hopefully have this ready for review/merge early next week.

noisysocks committed:
Make useBlockEditingMode use context

ramonjd commented on packages/block-editor/src/components/block-editing-mode/index.js:
```diff
@@ -0,0 +1,35 @@
+import { store as blockEditorStore } from '../../store';
+import { unlock } from '../../lock-unlock';
+
+export function useBlockEditingMode( clientId, mode ) {
```
Just confirming: we need a separate hook (as opposed to directly calling a `getBlockEditingMode` selector and `getBlockEditingMode` action in components) so we can clear the state when the component unmounts?

noisysocks replied:
There's no real benefit to using `useBlockEditingMode` instead of `getBlockEditingMode`. But yes the benefit of using `useBlockEditingMode` over `setBlockEditingMode` is that it handles setting and clearing the mode in the effect. Mostly this hook exists so that the API is as declarative for block authors as possible.

noisysocks committed:
Remove rootBlockEditingMode setting

ramonjd commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -18,3 +18,19 @@ export function isBlockInterfaceHidden( state ) {
+	) {
+		clientId = state.blocks.parents.get( clientId );
+	}
+	return (
```
I don't know how you feel about this, but do you think we should do any type defining for the `useBlockEditingMode` hook and associated selector and action? E.g.,

If we're dealing with an enum, it might help IDE users to know the values, and hopefully the ts compiler would catch any `blockEditingMode === 'ponies'` checks 🤷

noisysocks replied:
Good idea, I'll add a `@typedef` enum.

ramonjd reviewed (commented):
Great stuff. I've mainly added notes from last week. 

I like the direction (now I know a bit more about the problem, constraints and history, thank you!)

Happy to keep testing as it iterates.

noisysocks committed:
Fix private createRegistrySelector selectors

noisysocks committed:
Consolidate templateLock=contentOnly logic into getBlockEditingMode

noisysocks committed:
Hide disabled blocks from List View

noisysocks committed:
Hide disabled blocks from breadcrumbs

noisysocks committed:
Add doc comments

noisysocks committed:
Add unit tests

noisysocks committed:
Use @typedef to document mode param

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into try/block-editing-mode

noisysocks commented:
OK I'm happy with this now. I added some unit tests and inline docs, and fixed up everything @ramonjd and I talked about. The key thing that makes me happy is that the API is entirely private now so we're not tied down to anything.

Wanted to point out a few UI changes I've included here:

| Before | After |
| - | - |
| <img width="1624" alt="Screenshot 2023-05-23 at 12 10 08" src="https://github.com/WordPress/gutenberg/assets/612155/0e096414-225d-4831-9f01-285b051ecc67"> | <img width="1624" alt="Screenshot 2023-05-23 at 12 09 22" src="https://github.com/WordPress/gutenberg/assets/612155/bc3f4267-6de5-4f36-9e83-b87fbba1420a"> |

- List View will not display any blocks that have an editing mode of `'disabled'`. This means that if you look at a pattern with `templateLock=contentOnly` in the List View you'll see only its content children. Previously we wouldn't show children at all. This fixes part of https://github.com/WordPress/gutenberg/issues/50365.
- List View will not show the block settings menu if the editing mode is `'contentOnly'`. This makes the List View interface align with the visual interface.
- The Breadcrumbs component does not display any blocks that are `'disabled'`. This was another item that came up in https://github.com/WordPress/gutenberg/issues/50365.

noisysocks committed:
Restore packages/components/package.json from trunk

noisysocks commented on packages/data/src/redux-store/index.js:
```diff
@@ -221,12 +221,14 @@ export default function createReduxStore( key, options ) {
-										( state, ...args ) =>
-											selector( state.root, ...args )
-								),
+								mapValues( privateSelectors, ( selector ) => {
+									if ( selector.isRegistrySelector ) {
+										selector.registry = registry;
+									}
+
+									return ( state, ...args ) =>
+										selector( state.root, ...args );
+								} ),
```
Needed so that you can use `createRegistrySelector` in a private selector.

noisysocks committed:
Restore packages/block-library/src/post-title/edit.js from trunk

ramonjd commented:
Just giving it a manual test:

#### Patterns

I can't see any regression in `"templateLock":"contentOnly",` functionality. I created a simple pattern and the content is locked as expected. Text controls are reduced

<img width="802" alt="Screenshot 2023-05-23 at 2 20 57 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/9b683ea4-bce2-47af-adc7-71e52cdbe21f">

### Site editor

Page view works as expected, I can only edit blocks defined in `withBlockEditingMode`. To edit a template part I have to select it in the Template sidebar and click the edit button.
<img width="872" alt="Screenshot 2023-05-23 at 2 24 49 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/06cd247b-d8c6-4027-a416-736711a1be21">

When I select a template or template part however, can't seem to access blocks outside the list defined in `withBlockEditingMode`.

<img width="800" alt="Screenshot 2023-05-23 at 12 50 20 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/6eac2c38-9ebc-4774-a011-dd2d240a417e">

It seems to me it's an implementation detail and nothing to do with the approach in this PR.  I'm wondering if we need to check the context and set editing mode after we know we're dealing with a page id / post-content block in the side editor?

noisysocks committed:
Move BlockListBlockContext out of block.js so that it exists on mobile platforms

noisysocks commented on packages/block-editor/src/components/block-list/block-list-block-context.js:
```diff
@@ -0,0 +1,6 @@
+ */
+import { createContext } from '@wordpress/element';
+
+export const BlockListBlockContext = createContext( null );
```
Pulling this out to its own file so that it's available for use on both web and mobile (React native). Hopefully that fixes mobile tests.

noisysocks commented:
@ramonjd: Yeah those are problems to do with the testing code in this PR's description which is dumb and hacky. In this PR I'm just focused on the block-editor APIs that https://github.com/WordPress/gutenberg/pull/49980 will use. In the real implementation (https://github.com/WordPress/gutenberg/pull/49980) the Template panel will not be shown (we will sub it out for a Page panel) and we'll only enter this restricted editing mode when viewing a _page_ post type in the site editor.

noisysocks commented:
I've opened https://github.com/WordPress/gutenberg/pull/50857 which uses this API. Check it out if you'd like a more convenient way of testing these changes.

That said, I think the most important thing to test is that there are no regressions to the `templateLock=contentOnly` feature. Everything else is private so easy to fix and iterate on in follow-ups.

ramonjd commented:
I tested content locking in the site and post editor for the following blocks (which activate `templateLock` in their block.json files)

- navigation
- group
- cover
- columns
- column


Child content is locked as I expected with no regressions.

The only change I can see, which is an added bonus, is the addition of locked symbols on the child block items in list view.

| Trunk | This PR|
| ------------- | ------------- |
| <img width="300" alt="Screenshot 2023-05-23 at 7 45 03 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/0eb341a8-bbfc-4f96-9967-685f0d095797">  | <img width="300" alt="Screenshot 2023-05-23 at 7 37 51 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/a88798a0-dbfe-473d-9510-62aa9b9f1716"> |

ramonjd commented on packages/block-library/src/image/image.js:
```diff
@@ -158,7 +158,7 @@ export default function Image( {
 	const isResizable =
 		allowResize &&
-		! isContentLocked &&
+		blockEditingMode === 'default' &&
```
Optional: `blockEditingMode === 'default'` is used 3 times, we could cache in a const

ramonjd commented on packages/block-editor/src/store/test/private-selectors.js:
```diff
@@ -21,4 +25,186 @@ describe( 'private selectors', () => {
+				).toBe( mode );
+			} );
+
+			it( `should return ${ mode } if overriden by a parent`, () => {
```
```suggestion
			it( `should return ${ mode } if overridden by a parent`, () => {
```

ramonjd commented on packages/block-editor/src/store/test/private-selectors.js:
```diff
@@ -21,4 +25,186 @@ describe( 'private selectors', () => {
+					},
+				},
+			};
+			__experimentalHasContentRoleAttribute.mockReturnValueOnce( false );
```
I suppose this isn't necessary because `blockTypeName` is always going to be `undefined` (and `__experimentalHasContentRoleAttribute`  therefore will return `false`)

noisysocks replied:
True but I like the explicitness.

ramonjd replied:
You do you. 👍

ramonjd commented on packages/block-editor/src/components/block-list/use-in-between-inserter.js:
```diff
@@ -74,8 +76,12 @@ export function useInBetweenInserter() {
-				// Don't set the insertion point if the template is locked.
-				if ( getTemplateLock( rootClientId ) ) {
+				// Don't show the in-between inserter if the container block is
+				// locked or disabled.
```
The code is pretty clear. I'd vote to ditch this comment. No biggie.

ramonjd commented on packages/block-editor/src/components/block-editing-mode/index.js:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @return {BlockEditingMode} The current editing mode.
+ */
+export function useBlockEditingMode( mode ) {
```
Just thinking — you've probably thought about this — would there ever be a situation in which we'd want to set the block editing mode on a block programmatically, e.g., outside the block list context?

If so, would it be a future-proofing measure to pass an object to `useBlockEditingMode( { mode } )` so it'd be easier to add `clientId` later?

noisysocks replied:
My thinking was along the lines of the 80:20 rule and that that for more advanced uses, e.g. wanting to use the hook in `BlockToolbar` or somewhere else outside of the block hierarchy, you'd probably want to just use `getBlockEditingMode` and `setBlockEditingMode`.

ramonjd commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -8,3 +28,88 @@
+ */
+export const getBlockEditingMode = createRegistrySelector(
+	( select ) =>
+		( state, clientId = '' ) => {
```
You [mentioned](https://github.com/WordPress/gutenberg/pull/50643#issuecomment-1554060494) this selector might need to be memoized (?)

noisysocks replied:
Yes!

Unfortunately it's not possible to memoize a `createRegistrySelector` selector, but that's okay, because there are only two selectors that `getBlockEditingMode` calls that do a decent amount of computation (read: they have a `for` loop), and both of these are memoized. They are `getExplicitBlockEditingMode` and `__experimentalHasContentRoleAttribute`. In other words `getBlockEditingMode` is O(1) when called a second time with the same arguments.

ramonjd reviewed (approved):
Thanks for persisting with this one @noisysocks

noisysocks committed:
DRY up blockEditingMode check

noisysocks committed:
Fix typo

noisysocks committed:
Remove unnecessary comment

talldan commented:
@noisysocks Random thought, I wonder whether it might be worth consolidating visual/html mode into this API.

At the moment there's a very similar API for that (`getBlockMode` / `toggleBlockMode`).

I could imagine `default`, `contentOnly` and `disabled` are all considered visual (with default being the normal visual mode), and an `html` option exists for html mode.

bph commented:
>The key thing that makes me happy is that the API is entirely private now so we're not tied down to anything. (@noisysocks 

> If this gives us the right solution, in terms of behavior, then I'm sold. I like this API and I can even see ourselves making it public. (@youknowriad 

Curation / Governance is a big topic right now and this change caught my attention, as there is a need to add extensibility to parts of the site editor to enable more granular governance. It seems this API could be made public and delight a few agency and product developers. Maybe it's too early, hard for me to say, though. It might be worth a discussion, especially if this wouldn't be a hard lift. 
cc @mtias

Mamaduka commented on packages/block-editor/src/components/list-view/use-list-view-client-ids.js:
```diff
@@ -16,14 +17,30 @@ export default function useListViewClientIds( { blocks, rootClientId } ) {
+				getBlockEditingMode,
+			} = unlock( select( blockEditorStore ) );
+
+			const removeDisabledBlocks = ( tree ) => {
+				return tree.flatMap( ( { clientId, innerBlocks, ...rest } ) => {
+					if ( getBlockEditingMode( clientId ) === 'disabled' ) {
+						return removeDisabledBlocks( innerBlocks );
+					}
+					return [
+						{
+							clientId,
+							innerBlocks: removeDisabledBlocks( innerBlocks ),
+							...rest,
+						},
+					];
+				} );
+			};
```
This will return a new array reference on each call and cause the whole ListView to be rendered, even if the client IDs tree is the same.

Should we extract this into a new private memoized selector?

ramonjd replied:
I tested something out by adding the following selector to packages/block-editor/src/store/private-selectors.js.


```js
const EMPTY_ARRAY = [];
export const getEnabledBlocks = createSelector(
	( state, blocks = EMPTY_ARRAY ) =>
		blocks.flatMap( ( { clientId, innerBlocks, ...rest } ) => {
			if ( getBlockEditingMode( state, clientId ) === 'disabled' ) {
				return getEnabledBlocks( state, innerBlocks );
			}
			return [
				{
					clientId,
					innerBlocks: getEnabledBlocks( state, innerBlocks ),
					...rest,
				},
			];
		} ),
	( state ) => [ state.blockEditingModes, state.blocks.parents ]
);
```

Used in  packages/block-editor/src/components/list-view/use-list-view-client-ids.js

```js
...
			const {
				getDraggedBlockClientIds,
				getSelectedBlockClientIds,
				__unstableGetClientIdsTree,
				getEnabledBlocks,
			} = unlock( select( blockEditorStore ) );
...
				clientIdsTree: getEnabledBlocks(
					blocks ?? __unstableGetClientIdsTree( rootClientId )
				),
...
```

And it does prevent entire tree rerenders, which might be helpful for posts with lots of blocks:

## Before

https://github.com/WordPress/gutenberg/assets/6458278/eadef3c4-76d1-4ffc-9510-1b13ce264aee



## After

https://github.com/WordPress/gutenberg/assets/6458278/40e154ae-1aa6-45cb-b739-115c17195b94

I'm just not sure about the dependency array `[ state.blockEditingModes, state.blocks.parents ]`, which I copied from `getExplicitBlockEditingMode()` since that function is called in `getEnabledBlocks()` anyway.

Mamaduka commented on packages/block-editor/src/components/block-breadcrumb/index.js:
```diff
@@ -22,11 +23,18 @@ import { store as blockEditorStore } from '../../store';
 		const selectedBlockClientId = getSelectedBlockClientId();
 		return {
-			parents: getBlockParents( selectedBlockClientId ),
+			parents: getBlockParents( selectedBlockClientId ).filter(
+				( parentClientId ) =>
+					getBlockEditingMode( parentClientId ) !== 'disabled'
+			),
```
Same here.
